### PR TITLE
KPUS-4: implement generic push notification mechanism

### PR DESCRIPTION
**Pull request by danielfinke**
- add `endpoint_push_req` for sending generic push notifications
  - enables sending to an individual device or all push-enabled devices
    for a user
  - covers common push payload props between APNs/FCM
  - covers some select unique push payload props on APNs or FCM
- generic push payloads/headers using keypath map
- move `pm_*` module lookups to binaries to avoid exposing atom table
- convert legacy `push_req`s into `endpoint_push_req`s to avoid dual
  implementations
  - maintain compatibility features such as merging push app extra
    headers/envelope when building payloads

## Related PRs
- https://github.com/2600hz/kazoo-core/pull/555

### CODE REVIEW

**Comment by danielfinke** [comment]
@lazedo requesting your review on this, since I've made some changes to the incoming call notifications. To avoid having 2 different implementations in `pm_apple`/`pm_firebase`, I am converting `push_req`s into `endpoint_push_req`s before sending them to the modules.

File: src/pusher_listener.erl [hunk 1]
```diff
@@ -272,16 +293,171 @@ code_change(_OldVsn, State, _Extra) ->
...
+                    %% Backwards-compatibility for FCM - alert/sound were being
+                    %% nested under `"data"' instead of `"notifications"'
+                    kz_json:set_values([{[<<"alert">>, <<"loc-key">>], BodyKey}
+                                       ,{[<<"alert">>, <<"loc-args">>], BodyParams}
+                                       ,{<<"sound">>, Sound}
+                                       ], Data);
+                _ -> Data
+            end,
+
+    Definition = kapi_pusher:endpoint_push_req_definition(),
```
Comment (harenson) [inline src/pusher_listener.erl]: Use `kapi_pusher:api_definition(<<"endpoint_push_req">>)` instead.

File: src/pusher_listener.erl [hunk 2]
```diff
@@ -272,16 +293,171 @@ code_change(_OldVsn, State, _Extra) ->
...
+                              ]),
+
+    Data1 = case pm_module(TokenType) of
+                'pm_firebase' ->
+                    %% Backwards-compatibility for FCM - alert/sound were being
+                    %% nested under `"data"' instead of `"notifications"'
+                    kz_json:set_values([{[<<"alert">>, <<"loc-key">>], BodyKey}
+                                       ,{[<<"alert">>, <<"loc-args">>], BodyParams}
+                                       ,{<<"sound">>, Sound}
+                                       ], Data);
```
Comment (harenson) [inline src/pusher_listener.erl]: Improve readability by adding a new line here, to make it easier to see the 2 parameters' function call.

File: src/pusher_listener.erl [hunk 3]
```diff
@@ -272,16 +293,171 @@ code_change(_OldVsn, State, _Extra) ->
...
+
+    Definition = kapi_pusher:endpoint_push_req_definition(),
+    kz_json:set_values([{<<"Alert">>, Alert}
+                       ,{<<"Badge">>, Badge}
+                       ,{<<"Data">>, Data1}
+                       ,{<<"TTL">>, TTL}
+                       ,{<<"Sound">>, Sound}
+                       ,{<<"Event-Category">>, kapi_definition:category(Definition)}
+                       ,{<<"Event-Name">>, kapi_definition:name(Definition)}
+                       ], JObj1).
```
Comment (harenson) [inline src/pusher_listener.erl]: Improve readability by adding a new line here, to make it easier to see the 2 parameters' function call.

Comment (harenson) [inline src/pusher_listener.erl]: This function could have some tests ;)

File: src/pusher_module_util.erl [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    Acc1 = case kz_maps:get(GetKey, GetKeyMap) of
+               'undefined' -> Acc;
+
+               %% Field-specific setter functions
+               CF when is_function(CF, 2) -> CF(V, Acc);
+               CF when is_function(CF, 3) -> CF(V, JObj, Acc);
+
+               GetKey1 -> kz_maps:put(GetKey1, Acc, V)
+           end,
+    build_payload(FlatProps, JObj, GetKeyMap, Acc1).
```
Comment (harenson) [inline src/pusher_module_util.erl:66]: This function could have some tests ;)

**Review by harenson — CHANGES_REQUESTED** [review]

**Comment by danielfinke** [comment]
@harenson I added EUnit tests for both functions. They are failing in CI right now, though, because they depend on the `kazoo-core` changes.

**Pull request merged**